devito/arguments.py

```diff
@@ -158,198 +141,350 @@ class PtrArgument(Argument):
+        elif len(provided_values) == 1:
+            value = provided_values[0]
+        else:
+            value = None
+        return value
+
+    def visit_TensorArgument(self, o, param):
+        assert(isinstance(self.consumer, DimensionParameter))
+        return self.known_values[o].shape[param]
+
+
+class Dependency(object):
+    """ Object that represents an edge between two nodes on a dependency graph
+        Dependencies are directional, i.e. A -> B != B -> A
+        :param dependency_type: A dependency can either by of type "gets_value_from"
+               in which case this dependency will be followed for value derivation
+               or it can be of type "verified_by" in which case this dependency will be 
+               followed for verification. Both types of dependencies may exist between a
+               pair of nodes. However, only a single dependency of a type may exist
+               between any pair of nodes.
+        :param obj: Dependencies have a source node and a target node. The source node
+                    will store the dependency object on itself. The obj referred to here
+                    is the target node of the dependency. 
+        :param param: An optional parameter that might be required to evaluate the
+                      relationship being defined by this Dependency. e.g. when a Dimension
+                      derives its value from a SymbolicFunction's shape, this param carries
+                      the index of a dimension in the SymbolicFunction's shape. 
     """
+    GETS_VALUE_FROM = "gets_value_from"
+    VERIFIED_BY = "verified_by"
+    _types = ["gets_value_from", "verified_by"]
+    def __init__(self, dependency_type, obj, param=None):
+        assert(dependency_type in self._types)
+        self.dependency_type = dependency_type
+        self.obj = obj
+        self.param = param
 
-    @cached_property
-    def rtargs(self):
-        return (ScalarArgument(self.name, self, self.dtype),)
-
-
-class ArrayArgProvider(ArgumentProvider):
-
-    """ Class used to decorate Tensor Function objects with behaviour required for runtime
-        arguments.
-    """
+    def __repr__(self):
+        return "(" + self.dependency_type + ":" + str(self.obj) + ")"
 
-    @cached_property
-    def rtargs(self):
-        return (TensorArgument(self.name, self),)
 
+class UnevaluatedDependency(object):
+    def __init__(self, consumer, evaluator, extra_param=None):
+        self.consumer = consumer
+        self.evaluator = evaluator
+        self.extra_param = extra_param
 
-class ObjectArgProvider(ArgumentProvider):
+    def evaluate(self, known_values):
+        return self.evaluator(self.consumer, known_values, self.extra_param)
 
-    """ Class used to decorate Objects with behaviour required for runtime arguments.
-    """
 
-    @cached_property
-    def rtargs(self):
-        return (PtrArgument(self.name, self),)
+def runtime_arguments(parameters):
```

Comment:
Is this ever used?

Reply:
https://github.com/opesci/devito/pull/453/files#diff-1526813b3ca372b26a915777ac6e7f4fR251